File: golang/data-structure/queue/LinkedListQueue.go
package queue

import (
	"bytes"
	"errors"
	"fmt"
)

/**
 * @Author: Tomonori
 * @Date: 2020/2/20 18:47
 * @Title:
 * --- --- ---
 * @Desc:
 */
type LinkedListQueue struct {
	head *node
	tail *node
	size int
}

func NewLinkedListQueue() *LinkedListQueue {
	return &LinkedListQueue{
		head: nil,
		tail: nil,
	}
}

func (l *LinkedListQueue) GetSize() int {
	return l.size
}

func (l *LinkedListQueue) IsEmpty() bool {
	return l.size == 0
}

func (l *LinkedListQueue) Enqueue(t interface{}) error {
	if l.tail == nil {
		l.tail = newNode()
		l.head = l.tail
	} else {
		l.tail.next = newNodeWithVal(t)
		l.tail = l.tail.next
	}
	l.size++

	return nil
}

func (l *LinkedListQueue) Dequeue() (interface{}, error) {
	var err error
	if l.IsEmpty() {
		err = errors.New("Cannot dequeue from an empty queue")
	}

	deNode := l.head
	l.head = l.head.next
	deNode.next = nil

	if l.head == nil {
		l.tail = nil
	}
	l.size--

	return deNode, err
}

func (l *LinkedListQueue) GetFront() (interface{}, error) {
	var err error
	if l.IsEmpty() {
		err = errors.New("Cannot dequeue from an empty queue")
	}
	return l.head.val, err
}

func (l *LinkedListQueue) String() string {
	buffer := bytes.Buffer{}
	buffer.WriteString("Queue: ")
	buffer.WriteString("[head] ")

	current := l.head
	for current != nil {
		buffer.WriteString(fmt.Sprintf("%v -> ", current))
		current = current.next
	}
	buffer.WriteString(" nil [tail]")

	return buffer.String()
}

type node struct {
	val  interface{}
	next *node
}

func newNode() *node {
	return newNodeWithValAndNext(nil, nil)
}

func newNodeWithVal(val interface{}) *node {
	return newNodeWithValAndNext(val, nil)
}

func newNodeWithValAndNext(val interface{}, nde *node) *node {
	return &node{
		val:  val,
		next: nde,
	}
}

func (n *node) String() string {
	return fmt.Sprintf("%v", n.val)
}

File: java/data-structure/queue/src/main/java/ru/reimu/queue/LinkedListQueue.java
package ru.reimu.queue;

/**
 * @Author: Tomonori
 * @Date: 2020/2/20 18:15
 * @Title:
 * @Desc: ↓ ↓ ↓ ↓ ↓
 * -----
 */
public class LinkedListQueue<T> implements IQueue<T> {

    private Node head;
    private Node tail;
    private int size;

    public LinkedListQueue() {
        this.head = null;
        this.tail = null;
        this.size = 0;
    }

    @Override
    public int getSize() {
        return size;
    }

    @Override
    public boolean isEmpty() {
        return size == 0;
    }

    @Override
    public void enqueue(T t) {
        if (tail == null) {
            tail = new Node(t);
            head = tail;
        } else {
            tail.next = new Node(t);
            tail = tail.next;
        }
        size++;
    }

    @Override
    public T dequeue() {
        if (isEmpty()) {
            throw new IllegalArgumentException("Cannot dequeue from an empty queue.");
        }

        Node deNode = head;
        head = head.next;
        deNode.next = null;

        if (head == null) {
            tail = null;
        }
        size--;

        return deNode.t;
    }

    @Override
    public T getFront() {
        if (isEmpty()) {
            throw new IllegalArgumentException("Cannot dequeue from an empty queue.");
        }
        return head.t;
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        sb.append("Queue: head ");

        Node current = head;
        while (current != null) {
            sb.append(current + " -> ");
            current = current.next;
        }

        sb.append("null tail");

        return sb.toString();
    }

    private class Node {

        public T t;

        public Node next;

        public Node() {
            this(null, null);
        }

        public Node(T t) {
            this(t, null);
        }

        public Node(T t, Node next) {
            this.t = t;
            this.next = next;
        }

        @Override
        public String toString() {
            return t.toString();
        }
    }
}
